### DISCUSSION TITLE
Builtin functions in Zig

### DISCUSSION THREAD

**Discussion by rvoogdgeert** [helix-editor/helix#8480] [Troubleshooting]
When using a builtin function in Zig the editor removes the `@` after insertion. Try typing:

```
pub fun printName(self: @This()) void {}
```

When typing @ the menu opens and as soon as on chooses on of the options, e.g. `This`, the '@' is removed and doesn't return after enter.

**Comment by the-mikedavis** [comment]
This looks like a bug upstream in ZLS. Can you reproduce this with another editor?

ZLS sends us this completion item:

```json
{
  "detail": "@This() type",
  "documentation": {
    "kind": "markdown",
    "value": "Returns the innermost struct, enum, or union that this function call is inside. This can be useful for an anonymous struct that needs to refer to itself:\n\n```zig\nconst std = @import(\"std\");\nconst expect = std.testing.expect;\ntest \"@This()\" {\n    var items = [_]i32{ 1, 2, 3, 4 };\n    const list = List(i32){ .items = items[0..] };\n    try expect(list.length() == 4);\n}\nfn List(comptime T: type) type {\n    return struct {\n        const Self = @This();\n        items: []T,\n        fn length(self: Self) usize {\n            return self.items.len;\n        }\n    };\n}\n```\nWhen `@This()` is used at file scope, it returns a reference to the struct that corresponds to the current file."
  },
  "filterText": "This",
  "insertText": "This()",
  "insertTextFormat": 2,
  "kind": 3,
  "label": "@This",
  "sortText": "4_@This",
  "textEdit": {
    "newText": "This()",
    "range": {
      "end": {
        "character": 24,
        "line": 0
      },
      "start": {
        "character": 23,
        "line": 0
      }
    }
  }
},
```

For a document like so:

```zig
pub fn printName(self: @) void {}
```

with the cursor to the right of the `@`. The character between `start` and `end` is the `@` and ZLS is asking us to replace that range with just `This()` rather than `@This()`.

It looks like this only happens in function parameters. If I try a completion later like

```zig
const self = @This();
```

ZLS sends a range where `start` is equal to `end`, after the `@`, so the item is correctly completed.

**Reply by rvoogdgeert** [reply]
I tried the same in VSCode using this [extension](https://github.com/ziglang/vscode-zig) and in IntelliJ using [this one](https://plugins.jetbrains.com/plugin/18062-zig-support). Both of the editors behaved 'normally', i.e. they did not remove the `@`.
The behavior in in Helix, indeed, only occurs with function parameters.  Things like importing the standard library with `const std = @import("std");` work fine.
